File: Source/Alternet.UI.Pal/PropertyGridVariant.cpp
#include "PropertyGridVariant.h"

namespace Alternet::UI
{
	/*static*/ void PropertyGridVariant::Delete(void* handle)
	{
		delete (PropertyGridVariant*)handle;
	}

	/*static*/ void* PropertyGridVariant::CreateVariant()
	{
		return new PropertyGridVariant();
	}

	PropertyGridVariant::PropertyGridVariant(wxVariant value)
	{
		variant = value;
	}

	PropertyGridVariant::PropertyGridVariant()
	{
	}
	
	PropertyGridVariant::~PropertyGridVariant()
	{
	}

	wxVariant& PropertyGridVariant::ToVar(void* handle)
	{
		return ((PropertyGridVariant*)handle)->variant;
	}

	void PropertyGridVariant::FromVariant(void* handle, wxVariant& value)
	{
		((PropertyGridVariant*)handle)->variant = value;
	}

	bool PropertyGridVariant::IsNull(void* handle)
	{
		return ToVar(handle).IsNull();
	}

	bool PropertyGridVariant::Unshare(void* handle)
	{
		return ToVar(handle).Unshare();
	}

	string PropertyGridVariant::GetValueType(void* handle)
	{
		return wxStr(ToVar(handle).GetType());
	}

	bool PropertyGridVariant::IsType(void* handle, const string& type)
	{
		return ToVar(handle).IsType(wxStr(type));
	}

	string PropertyGridVariant::MakeString(void* handle) 
	{
		return wxStr(ToVar(handle).MakeString());
	}

	Color PropertyGridVariant::GetColor(void* handle)
	{
		wxVariant value = ToVar(handle);
		if (value.IsNull())
			return Color();
		wxColour result;
		auto vtype = value.GetType();
		if (vtype == "wxColour")
		{
			result << value;
			return result;
		}
		if (vtype == "wxColourPropertyValue")
		{
			wxAny any = value;
			wxColourPropertyValue cpv = any.As<wxColourPropertyValue>();
			result = cpv.m_colour;
			return Color(result, cpv.m_type);
		}
		return Color();
	}

	void PropertyGridVariant::SetColor(void* handle, const Color& val) 
	{
		wxColor color = val;
		wxVariant v;
		v << color;
		FromVariant(handle, v);
	}

	void PropertyGridVariant::MakeNull(void* handle)
	{
		wxVariant v = wxVariant();
		FromVariant(handle, v);
	}

	void PropertyGridVariant::Clear(void* handle)
	{
		wxVariant v = wxVariant();
		FromVariant(handle, v);
	}

	double PropertyGridVariant::GetDouble(void* handle) 
	{
		return ToVar(handle).GetDouble();
	}

	bool PropertyGridVariant::GetBool(void* handle) 
	{
		return ToVar(handle).GetBool();
	}

	int64_t PropertyGridVariant::GetLong(void* handle) 
	{
		return ToVar(handle).GetLongLong().GetValue();
	}
	
	uint64_t PropertyGridVariant::GetULong(void* handle)
	{
		return ToVar(handle).GetULongLong().GetValue();
	}

	DateTime PropertyGridVariant::GetDateTime(void* handle) 
	{
		wxVariant value = ToVar(handle);
		if (value.IsNull())
			return DateTime();

		wxDateTime dt = value.GetDateTime();
		if(dt.IsValid())
			return dt;
		DateTime dt2 = DateTime();
		dt2.Reset();
		return dt2;
	}

	string PropertyGridVariant::GetString(void* handle) 
	{
		return wxStr(ToVar(handle).GetString());
	}

	void PropertyGridVariant::SetDouble(void* handle, double val) 
	{
		wxVariant v = val;
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetBool(void* handle, bool val) 
	{
		wxVariant v = val;
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetUInt(void* handle, uint32_t val)
	{
		wxAny any = val;
		wxVariant v = any;
		FromVariant(handle, v);
	}

	int PropertyGridVariant::GetInt(void* handle)
	{
		return ToVar(handle).GetInteger();
	}

	uint32_t PropertyGridVariant::GetUInt(void* handle)
	{
		return (uint32_t)ToVar(handle).GetInteger();
	}

	void PropertyGridVariant::SetULong(void* handle, uint64_t val)
	{
		wxVariant v = wxULongLong(val);
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetLong(void* handle, int64_t val)
	{
		wxVariant v = wxLongLong(val);
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetInt(void* handle, int val) 
	{
		wxVariant v = val;
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetShort(void* handle, int16_t val) 
	{
		wxVariant v = val;
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetDateTime(void* handle, const DateTime& val) 
	{
		wxVariant v = wxDateTime(val);
		FromVariant(handle, v);
	}

	void PropertyGridVariant::SetString(void* handle, const string& value) 
	{
		wxVariant v = wxStr(value);
		FromVariant(handle, v);
	}
}

File: Source/NativeApi/NativeApi/Api/PropertyGrid/PropertyGridVariant.cs
#pragma warning disable
using ApiCommon;
using System;
using System.Collections.Generic;
using System.Text;
using Alternet.Drawing;

namespace NativeApi.Api
{
    //https://docs.wxwidgets.org/3.2/classwx_variant.html
    public class PropertyGridVariant
    {
        public static void Delete(IntPtr handle) { }

        public static IntPtr CreateVariant() => throw new Exception();

        public static bool IsNull(IntPtr handle) => throw new Exception();
        public static bool Unshare(IntPtr handle) => throw new Exception();
        public static void MakeNull(IntPtr handle) => throw new Exception();
        public static void Clear(IntPtr handle) => throw new Exception();
        public static string GetValueType(IntPtr handle) => throw new Exception();
        public static bool IsType(IntPtr handle, string type) => throw new Exception();
        public static string MakeString(IntPtr handle) => throw new Exception();

        public static Color GetColor(IntPtr handle) => throw new Exception();
        public static double GetDouble(IntPtr handle) => throw new Exception();
        public static bool GetBool(IntPtr handle) => throw new Exception();

        public static int GetInt(IntPtr handle) => throw new Exception();
        public static uint GetUInt(IntPtr handle) => throw new Exception();

        public static long GetLong(IntPtr handle) => throw new Exception();
        public static ulong GetULong(IntPtr handle) => throw new Exception();
        public static Alternet.UI.DateTime GetDateTime(IntPtr handle) => throw new Exception();
        public static string GetString(IntPtr handle) => throw new Exception();

        public static void SetColor(IntPtr handle, Color val) => throw new Exception();
        public static void SetDouble(IntPtr handle, double val) => throw new Exception();
        public static void SetBool(IntPtr handle, bool val) => throw new Exception();
        public static void SetLong(IntPtr handle, long val) => throw new Exception();
        public static void SetULong(IntPtr handle, ulong val) => throw new Exception();
        public static void SetInt(IntPtr handle, int val) => throw new Exception();
        public static void SetUInt(IntPtr handle, uint val) => throw new Exception();
        public static void SetShort(IntPtr handle, short val) => throw new Exception();
        public static void SetDateTime(IntPtr handle, Alternet.UI.DateTime val) =>
            throw new Exception();
        public static void SetString(IntPtr handle, string value) => throw new Exception();
    }
}